bitcoin/src/blockdata/fee_rate.rs

```diff
@@ -82,10 +82,12 @@ impl FeeRate {
 
     /// Checked weight multiplication.
     ///
-    /// Computes `self * rhs` where rhs is of type Weight. `None` is returned if an overflow
-    /// occurred.
+    /// Computes the absolute fee amount for a given [`Weight`] at this fee rate.
+    ///
+    /// `None` is returned if an overflow occurred.
     pub fn checked_mul_by_weight(self, rhs: Weight) -> Option<Amount> {
-        self.0.checked_mul(rhs.to_wu()).map(Amount::from_sat)
+        let sats = self.0.checked_mul(rhs.to_wu())?.checked_add(999)? / 1000;
```

Comment:
good catch, but wouldn't this be simpler read using checked_mul instead of checked_add?

suggest
```
let sats = self.0.checked_mul(rhs.to_wu())?.checked_mul(1000)?;
```

I don't really understand why the checked_add here.

Reply:
Ah, alright, this is confusing.  It feels like what we really want is for `rhs: Weight` to instead be `KWU` instead of WU.  Right?

Reply:
In other-words, if fee_rate is in sats/kwu, then wouldn't the interface be simpler as:  then sats/kwu * kwu = sats

```
    pub fn checked_mul_by_weight(self, rhs: KWU) -> Option<Amount> {
        let sats = self.0.checked_mul(rhs)?;
        Some(Amount::from_sat(sats))
    }
```

Reply:
> It feels like what we really want is for rhs: Weight to instead be KWU instead of WU. Right?

Yes i suppose that's a good way to visualize it, but we couldn't implement it that way. We'd have to represent `KWU` as a float (or a new struct type) to perform accurate arithmetic on weight-unit numbers less than 1000.

Reply:
> We'd have to represent KWU as a float (or a new struct type)

Yes, exactly, like here:
https://github.com/rust-bitcoin/rust-bitcoin/pull/1787/files#diff-c2ea40075e93ccd068673873166cfa3312ec7439d6bc5a4cbc03e972c7e045c4R16

Reply:
Looks like #1787 also fixes the same bug as this PR. Although it's much bigger and has been held up since April so maybe a small patch like this PR might be appropriate until #1787 can be merged. I filed this PR because I got unexpected results back from `FeeRate::fee_wu`. #1787 is probably a better place to consider refactoring APIs; this PR is just a hotfix after all.

Reply:
Definitely let's get this fix in here. No point waiting for #1787 

I'd be in favor of making a point release as well.

Reply:
Yeah definitively better to not wait for #1787